url47fcfc0133_baliosian_tau_fst_tgz-pJ8-uy_edu_fing_mina_fsa_test_old_TestComposition1J8/src/uy/edu/fing/mina/lupa/LupaExporter.java

```diff
@@ -0,0 +1,826 @@
+	 */
+	private static String updateMainActionsInLupa(String old, Tffst tffst) {
+//		Iterator<Transition> transIt = tffst.getTransitions().iterator();
+		String out = new String(old);
+		
+//		Set<String> alreadyGenerated = new HashSet<String>();
+////		We take only those functions that are Events.
+//		List<TfI> workingList = new LinkedList<TfI>();
+////		TODO Why are we working only with the first element of the LabelIn?
+//		TfI firstElementOfTfString;
+//		String functionName;
+//		for(Transition t : tffst.getTransitions()){
+//			
+//			if(!t.getLabelIn().isEpsilon()){
+//				firstElementOfTfString = t.getLabelIn().get(0);
+//				if(!firstElementOfTfString.isNot() && (firstElementOfTfString instanceof EventTf)){
+//					workingList.add(firstElementOfTfString);
+//				}
+//			}
+//		}
+		
+		String functionHeader;
+		for(String functionName : mainActions.keySet()){
+			functionHeader = functionHeader(functionName);
+//			There are no duplicates in a Map.
+			if(! out.contains(functionHeader)){
+				out = out.concat("\t-- " + functionHeader + "\n");
+				out = out.concat("actions." + functionName + " = function(e)\n");
+//				TODO Mejorar!!
+				
+				out = out.concat("\te = shared[\"incomming_event\"]\n");
+				out = out.concat("\t------------------------------------------------\n");
+				out = out.concat("\t-- TODO: Complete this with your action code. --\n");
+				out = out.concat("\t------------------------------------------------\n");
+				out = out.concat("end\n\n");
+			}
+			
+		}
+		return out;
+	}
+
+	/**
+	 * Creates a String with event functions that are CompositeTf or Nots.
+	 * @param tffst The TFFST to export.
+	 * @return The updated String objects with the new function bodies that it may have.
+	 * @throws UnsupportedTFFSTException 
+	 */
+	private static String compositeEventsToLupa(Tffst tffst) throws UnsupportedTFFSTException {
+		String out = new String();
+		
+		out = out.concat("-----------------------------------------------\n");
+		out = out.concat("--          BEGIN COMPOSITE EVENTS           --\n");
+		out = out.concat("-----------------------------------------------\n");
+		CompositeTf comp;
+		TfI leftTf;
+		TfI rightTf;
+		String functionName;
+		for(TfI function : compositeEvents.values()){
+			functionName = functionName(function);
+			if(function instanceof CompositeTf){
+				comp = (CompositeTf) function;
+				leftTf = comp.left;
+				rightTf = comp.right;
+				out = out.concat("-- " + function.getName() + "\n");
+				out = out.concat("events." + functionName + " = function(e) \n");
+//				out = out.concat(functionHeader(functionName) + "\n");
+				out = out.concat("\tlocal left= events." + functionName(leftTf) + "(e)\n");
+				out = out.concat("\tlocal right= events."	+ functionName(rightTf) + "(e)\n");
+				if (comp.op == Operator.AND) {
+					out = out.concat("\tif left < right then return left else return right end\n");
+				} else if (comp.op == Operator.OR) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An equality operator (== or !=) is used to compare strings, which matches strings based on address.

**Fix**: Do content comparison using the equals() method to compare the values of the strings.